### PR TITLE
TMEDIA-851 : Component - Badge

### PR DESCRIPTION
**Pull request by ecarlisle**
## Ticket

- [TMEDIA-851 : Component - Badge](https://arcpublishing.atlassian.net/browse/TMEDIA-851)

## Description
1. Creates the Badge component, based on its [Figma design](https://www.figma.com/file/vWVGgZfZVBdSX7ACo3kmjr/Enhanced-Styling-Design-System?node-id=3505%3A27776).
2. Only nonstandard property is `variant`, corresponding with the different designs of the Badge.
3. **Note:** Like `<Pill>`, `<Badge>` uses `border-radius: 2em` until a token for rounded left- and right-sides and standardized with a token.
4. **Note:** This PR branches from the `TMEDIA-788-style-commerce-theme` branch to be able to use global and alias available in that branch.

## Acceptance Criteria
- Provide a `<Badge>` component 
    - Has the follow variants
        - default
        - light
        - primary
        - danger
        - warning
        - success
    - Have Scss file that adds the hooks to the Sass library
    - Have test coverage
    - Have storybook documentation
        - Note: Until we have a theme switcher, there won’t be a background to see the styles

## Test Steps

1. Checkout branch - `git checkout TMEDIA-851-component-badge`
2. Update dependencies - `npm i`
5. Run Storybook `npm run storybook`
6. Confirm requirements by viewing Badge in Storybook using the Commerce theme.

## Author Checklist

_The author of the PR should fill out the following sections to ensure this PR is ready for review._

- [ ] Confirmed all the test steps a reviewer will follow above are working.
- [ ] Ran this code locally and checked that there are not any unintended side effects. For example, that a CSS selector is scoped only to a particular block.
- [ ] Confirmed relevant documentation has been updated/added.
- [ ] Add label - **ready for review** when the pull request is ready for someone to begin reviewing

## Reviewer Checklist

_The reviewer of the PR should copy-paste this template into the review comments on review._

- [ ] All GitHub Actions pass
- [ ] Ran the code locally based on the test instructions.
- [ ] Checked Chromatic for Storybook changes, accepted the updates if acceptable
- [ ] Looked to see that the new or changed code has code coverage, specifically. We want the global code coverage to keep on going up with targeted testing.
- [ ] Approve and Add label - **ready to merge** if you are happy with the pull request
- [ ] Want another reviewer? Add the label **additional review**

### CODE REVIEW

File: .storybook/themes/commerce.scss [hunk 1]
```diff
@@ -217,6 +217,32 @@
...
+				badge: (
+					border-radius: 2em,
+					color: var(--global-white),
+					font-size: var(--body-font-size-tiny),
+					font-weight: var(--global-font-weight-7),
+					line-height: var(--body-line-height-tiny),
+					padding: var(--global-spacing-1) var(--global-spacing-2),
+				),
+				badge-danger: (
+					background-color: var(--form-border-color-danger),
```
Comment (matthewroach) [inline .storybook/themes/commerce.scss:229]: We should not be using `form` tokens or `border` tokens here - we have more relevant tokens such as `background-color-danger`

File: .storybook/themes/commerce.scss [hunk 2]
```diff
@@ -217,6 +217,32 @@
...
+				),
+				badge-default: (
+					background-color: var(--text-color),
```
Comment (matthewroach) [inline .storybook/themes/commerce.scss:232]: Using a text token for background is not correct

File: .storybook/themes/commerce.scss [hunk 3]
```diff
@@ -217,6 +217,32 @@
...
+				),
+				badge-light: (
+					background-color: var(--text-color-subtle),
```
Comment (matthewroach) [inline .storybook/themes/commerce.scss:235]: Using a text token for background is not correct

File: .storybook/themes/commerce.scss [hunk 4]
```diff
@@ -217,6 +217,32 @@
...
+				),
+				badge-primary: (
+					background-color: var(--color-primary),
+				),
+				badge-success: (
+					background-color: var(--form-border-color-success),
```
Comment (matthewroach) [inline .storybook/themes/commerce.scss:241]: Should be `background-color-success`

File: .storybook/themes/commerce.scss [hunk 5]
```diff
@@ -217,6 +217,32 @@
...
+				),
+				badge-warning: (
+					background-color: var(--form-border-color-warning),
```
Comment (matthewroach) [inline .storybook/themes/commerce.scss:244]: Should be `background-color-warning`

File: src/components/badge/index.test.jsx [hunk 6]
```diff
@@ -0,0 +1,40 @@
+import { render, screen } from "@testing-library/react";
+
+import Badge from ".";
+
+const BADGE_TEXT = "Hello World";
+describe("Badge", () => {
+	it("Should render string child.", () => {
+		render(<Badge>{BADGE_TEXT}</Badge>);
+		expect(screen.queryByText("Hello World")).not.toBeNull();
```
Comment (matthewroach) [inline src/components/badge/index.test.jsx]:
```suggestion
		expect(screen.queryByText(BADGE_TEXT)).not.toBeNull();
```

File: src/components/badge/index.test.jsx [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+	it("should allow pass through of props", () => {
+		render(<Badge id="custom-id">{BADGE_TEXT}</Badge>);
+		expect(screen.queryByText(BADGE_TEXT)).toHaveAttribute("id", "custom-id");
+	});
+
+	it("should render additional classes", () => {
+		const ORIGINAL_CLASSES = "c-badge";
+		const ADDITIONAL_CLASSES = "additionalClass1 additionalClass2";
+		render(<Badge className={ADDITIONAL_CLASSES}>Hello World</Badge>);
+		const element = screen.queryByText("Hello World");
```
Comment (matthewroach) [inline src/components/badge/index.test.jsx]:
```suggestion
		const element = screen.queryByText(BADGE_TEXT);
```

File: src/components/badge/index.test.jsx [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+
+	it("should allow pass through of props", () => {
+		render(<Badge id="custom-id">{BADGE_TEXT}</Badge>);
+		expect(screen.queryByText(BADGE_TEXT)).toHaveAttribute("id", "custom-id");
+	});
+
+	it("should render additional classes", () => {
+		const ORIGINAL_CLASSES = "c-badge";
+		const ADDITIONAL_CLASSES = "additionalClass1 additionalClass2";
+		render(<Badge className={ADDITIONAL_CLASSES}>Hello World</Badge>);
```
Comment (matthewroach) [inline src/components/badge/index.test.jsx]:
```suggestion
		render(<Badge className={ADDITIONAL_CLASSES}>{BADGE_TEXT}</Badge>);
```

File: src/components/badge/index.test.jsx [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+		const element = screen.queryByText("Hello World");
+		expect(element).toHaveClass(ADDITIONAL_CLASSES);
+		expect(element).toHaveClass(ORIGINAL_CLASSES);
+	});
+
+	it("should use the 'default' className when no variant is indicated.", () => {
+		const DEFAULT_CLASS = "c-badge--default";
+		render(<Badge>{BADGE_TEXT}</Badge>);
+		const element = screen.queryByText("Hello World");
```
Comment (matthewroach) [inline src/components/badge/index.test.jsx]:
```suggestion
		const element = screen.queryByText(BADGE_TEXT);
```

File: src/components/badge/index.test.jsx [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+		expect(element).toHaveClass(DEFAULT_CLASS);
+	});
+
+	it("should use a variant's respective className.", () => {
+		const variant = "success";
+		const SUCCESS_CLASS = "c-badge--success";
+		render(<Badge variant={variant}>{BADGE_TEXT}</Badge>);
+		const element = screen.queryByText("Hello World");
```
Comment (matthewroach) [inline src/components/badge/index.test.jsx]:
```suggestion
		const element = screen.queryByText(BADGE_TEXT);
```

**Review by matthewroach — CHANGES_REQUESTED** [review]

File: .storybook/themes/commerce.scss (hunk 1 above)
Comment (ecarlisle) [reply .storybook/themes/commerce.scss:229]: Agree 100% with this and other similar comments with the badge background colors. I was going w/ the current design in Figma, and there is a discussion on that card.
image
I'll check w/ Design as we don't currently have all of these colors under Status Colors in Figma. 
image

Comment (matthewroach) [reply .storybook/themes/commerce.scss:229]: Created https://arcpublishing.atlassian.net/browse/TMEDIA-857 to address the tokens in use once design have finalised them

**Review by matthewroach — APPROVED** [review]

**Review by JackHowa — APPROVED** [review]
Works as expected. Good work. lgtm

**Pull request merged**